### PR TITLE
[SPARK-7970] Skip closure cleaning for SQL operations

### PR DESCRIPTION
**Pull request by nitin2goyal**
Also introduces new spark private API in RDD.scala with name 'mapPartitionsInternal' which doesn't closure cleans the RDD elements.

### CODE REVIEW

**Comment by nitin2goyal** [comment]
cc @andrewor14

**Comment by andrewor14** [comment]
ok to test

**Comment by nitin2goyal** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -706,6 +706,23 @@ abstract class RDD[T: ClassTag](
   }
 
   /**
+   * Spark's internal mapPartitions method which skips closure cleaning. To be used carefully
+   * only if we are sure that the RDD elements are serializable and don't require closure
+   * cleaning
+   *
+   * `preservesPartitioning` indicates whether the input function preserves the partitioner, which
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: just use `@param` here

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 2]
```diff
@@ -706,6 +706,23 @@ abstract class RDD[T: ClassTag](
   }
 
   /**
+   * Spark's internal mapPartitions method which skips closure cleaning. To be used carefully
+   * only if we are sure that the RDD elements are serializable and don't require closure
+   * cleaning
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: can you add that this is mainly for performance improvements? Also you're missing a period at the end.

**Comment by andrewor14** [comment]
Looks great! I look forward to getting this merged. Once you address the comments I will do so.

**Comment by nitin2goyal** [comment]
Thanks fore reviewing Andrew ( @andrewor14 ). Have addressed your comments. Let me know if it looks good.

**Comment by andrewor14** [comment]
@nitin2goyal Sorry for the delay. This LGTM. I will merge it once you rebase to master again.

**Comment by andrewor14** [comment]
retest this please

**Comment by tedyu** [comment]
Should mapPartitions() be replaced with mapPartitionsInternal() in the following classes ?
```
    child.execute().mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala
    val rootType = schemaData.mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala
    json.mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala
    rows.mapPartitions { iterator =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONRelation.scala
        .mapPartitions { iterator =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala
      .mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/DefaultSource.scala
      child.execute().mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala
    data.mapPartitions { iterator =>
    data.mapPartitions { iterator =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala
    child.execute().mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/Expand.scala
    streamedPlan.execute().mapPartitions { streamedIter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala
    streamedPlan.execute().mapPartitions { streamedIter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashOuterJoin.scala
    val matchesOrStreamedRowsWithNulls = streamed.execute().mapPartitions { streamedIter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoin.scala
    streamed.execute().mapPartitions { streamedIter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/LeftSemiJoinBNL.scala
    rdd.mapPartitions { iter =>
    inputRDD.mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/python.scala
    child.execute().mapPartitions { iter =>
    child.execute().mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/rowFormatConverters.scala
    child.execute().mapPartitions { iter =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/sort.scala
    child.execute().mapPartitions { stream =>
/Users/tyu/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala
```
If so, allow me to open a PR

**Pull request closed without merging**